presto-main/src/main/java/com/facebook/presto/server/protocol/QueryBlockingRateLimiter.java

```diff
@@ -0,0 +1,107 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.server.protocol;
+
+import com.facebook.airlift.stats.CounterStat;
+import com.facebook.presto.execution.QueryManagerConfig;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.inject.Inject;
+import org.weakref.jmx.Managed;
+import org.weakref.jmx.Nested;
+
+import javax.annotation.PreDestroy;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.util.Objects.requireNonNull;
+
+/*
+ * Rate Limiting per query with token bucket
+ * Rate = rateLimitBucketMaxSize/second
+ * When having sufficient tokens, Request will be responded immediately.
+ * When not having enough tokens available, it uses the delayed processing method.
+ */
+public class QueryBlockingRateLimiter
+{
+    private final LoadingCache<QueryId, RateLimiter> rateLimiterCache;
+    private final long rateLimitBucketMaxSize;
+    private final ListeningExecutorService rateLimiterExecutorService;
+
+    private CounterStat rateLimiterTriggeredCounter = new CounterStat();
+
+    @Inject
+    public QueryBlockingRateLimiter(QueryManagerConfig queryManagerConfig)
+    {
+        requireNonNull(queryManagerConfig, "queryManagerConfig is null");
+        this.rateLimitBucketMaxSize = queryManagerConfig.getRateLimitBucketMaxSize();
+        // Using a custom thread pool with size 1-10 to reduce initial thread resources
+        ExecutorService executorService = new ThreadPoolExecutor(1, 10, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(), daemonThreadsNamed("rate-limiter-listener"));
+        rateLimiterExecutorService = listeningDecorator(executorService);
+        rateLimiterCache = CacheBuilder.newBuilder().maximumSize(10000).expireAfterAccess(60, TimeUnit.MINUTES).build(CacheLoader.from(key -> RateLimiter.create(rateLimitBucketMaxSize)));
```

Comment:
this one was using the same default value as gateway.
I was also debating if if is worth it to add all of them as configurations such as maximumSize and expiration, etc.

Reply:
The reason I'm asking is because if we have a surge of traffic coming in (eg: 1M request/s), with the unlimited queue size, wouldn't it be a non-trivial amount of overhead as well?

So just in case in the future we hit such issue in production, we should be able to tune the executor pool for rate limiter instead of waiting for another release

Reply:
Coordinator Rate limiting is on each query level, so it will throttle for that problematic query's traffic. 
Do you mean, If there are 1M different queries hitting the same coordinator, the overhead will be non-trivial?